### PR TITLE
Rpcdaemon: fixed few bugs in erigon_getLatestLogs

### PR DESCRIPTION
**Pull request by lupin012**
doing some testing some bugs in the API implementation were found and fixed

### CODE REVIEW

File: turbo/jsonrpc/erigon_receipts.go [hunk 1]
```diff
@@ -218,26 +218,47 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
...
-		latest, err = rpchelper.GetLatestBlockNumber(tx)
-		//to fetch latest
-		latest += 1
+		// Convert the RPC block numbers into internal representations
+		latest, err := rpchelper.GetLatestBlockNumber(tx)
 		if err != nil {
 			return nil, err
 		}
+
+		begin = 0
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/erigon_receipts.go:238]: all this logic need to be inside `applyFilters` func

File: turbo/jsonrpc/erigon_receipts.go [hunk 2]
```diff
@@ -301,9 +322,10 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
 				blockLogs = append(blockLogs, filtered[i])
 				logCount++
 			}
-			if logOptions.LogCount != 0 && logOptions.LogCount == logCount {
-				continue
+			if logOptions.LogCount != 0 && logOptions.LogCount <= logCount {
+				break
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/erigon_receipts.go:326]: why break only internal loop? maybe need break both?

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: turbo/jsonrpc/erigon_receipts.go [hunk 3]
```diff
@@ -218,26 +218,47 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
...
 	var err error
-	if crit.ToBlock != nil {
-		if crit.ToBlock.Sign() >= 0 {
-			latest = crit.ToBlock.Uint64()
-		} else if !crit.ToBlock.IsInt64() || crit.ToBlock.Int64() != int64(rpc.LatestBlockNumber) {
-			return nil, fmt.Errorf("negative value for ToBlock: %v", crit.ToBlock)
+	var begin, end uint64
+
+	if crit.BlockHash != nil {
+		number := rawdb.ReadHeaderNumber(tx, *crit.BlockHash)
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/erigon_receipts.go:225]: please use _blockReader

File: turbo/jsonrpc/erigon_receipts.go [hunk 4]
```diff
@@ -218,26 +218,47 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
...
 	}
 	defer tx.Rollback()
-	var latest uint64
 	var err error
-	if crit.ToBlock != nil {
-		if crit.ToBlock.Sign() >= 0 {
-			latest = crit.ToBlock.Uint64()
-		} else if !crit.ToBlock.IsInt64() || crit.ToBlock.Int64() != int64(rpc.LatestBlockNumber) {
-			return nil, fmt.Errorf("negative value for ToBlock: %v", crit.ToBlock)
+	var begin, end uint64
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/erigon_receipts.go:222]: please add comment about semantic of his variables. is it `[begin, end)`?

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

**Comment by AskAlexSharov** [comment]
FYI: after review is done - please apply same logic for erigon_getLogs and getLogsV3(if need)

File: turbo/jsonrpc/erigon_receipts.go [hunk 5]
```diff
@@ -218,26 +218,47 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
...
+		number := rawdb.ReadHeaderNumber(tx, *crit.BlockHash)
+		if number == nil {
+			return nil, fmt.Errorf("block not found: %x", *crit.BlockHash)
 		}
+		begin = *number
+		end = *number
 	} else {
-		latest, err = rpchelper.GetLatestBlockNumber(tx)
-		//to fetch latest
-		latest += 1
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/erigon_receipts.go:232]: if semantic is `[begin, end)` - maybe += 1 still needed

File: turbo/jsonrpc/erigon_receipts.go (hunk 2 above)
Comment (AskAlexSharov) [reply turbo/jsonrpc/erigon_receipts.go:329]: here is 2 loops:
```
for iter.HasNext() {
it, err := tx.Prefix(kv.Log, hexutility.EncodeTs(blockNumber))
		if err != nil {
			return nil, err
		}
		for it.HasNext() {
                       // your break is here, and will break only internal loop
```

Comment (lupin012) [reply turbo/jsonrpc/erigon_receipts.go:329]: here is 2 loops:

for iter.HasNext() {
                it, err := tx.Prefix(kv.Log, hexutility.EncodeTs(blockNumber))
		if err != nil {
			return nil, err
		}
		for it.HasNext() {
                       fill blockLogs list
                       if logOptions.LogCount != 0 && logOptions.LogCount <= logCount {
                                break
                 }
                 
                 fills erigonLog from data of blockLogs and others
                 
                  if logOptions.LogCount != 0 && logOptions.LogCount <= logCount {
                        return erigonLogs, nil
                }
                if logOptions.BlockCount != 0 && logOptions.BlockCount <= blockCount {
                        return erigonLogs, nil
                }

      }

Comment (AskAlexSharov) [reply turbo/jsonrpc/erigon_receipts.go:329]: Ok

**Review by AskAlexSharov — APPROVED** [review]

**Pull request merged**